### PR TITLE
[App] Add support for plugins to return actions

### PR DESCRIPTION
**Pull request by ethanwharris**
## What does this PR do?

- Allows a plugin to return actions that show a toast message or navigate to a page.
- Refactors LightningPlugin to live in `lightning.app.plugin`

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @borda

### CODE REVIEW

**Comment by awaelchli** [comment]
@ethanwharris I think the `tests/__init__.py` was added by accident. I think it should be removed, this could be the cause of the import errors in the tests.

File: src/lightning/app/utilities/cloud.py [hunk 1]
```diff
@@ -22,10 +23,11 @@
 from lightning.app.utilities.network import LightningClient
 
 
-def _get_project(
-    client: LightningClient, project_id: str = LIGHTNING_CLOUD_PROJECT_ID, verbose: bool = True
-) -> V1Membership:
+def _get_project(client: LightningClient, project_id: Optional[str] = None, verbose: bool = True) -> V1Membership:
     """Get a project membership for the user from the backend."""
+    if project_id is None:
+        project_id = LIGHTNING_CLOUD_PROJECT_ID
```
Comment (awaelchli) [inline src/lightning/app/utilities/cloud.py:29]: I'm wondering what the reason is. Is LIGHTNING_CLOUD_PROJECT_ID not a constant, being modified somewhere?

**Review by awaelchli — APPROVED** [review]

**Comment by carmocca** [comment]
I created a PR for the 1.9.4 release. @ethanwharris Can you cherry-pick this changes into https://github.com/Lightning-AI/lightning/pull/16884?

**Comment by carmocca** [comment]
Should this have a CHANGELOG entry?

**Comment by ethanwharris** [comment]
@carmocca No, not user facing at this time 😃

**Pull request merged**